src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,594 @@
+    A class could have methods, subclasses referencing other objects.
+
+    Args:
+        cls: The class to be analyzed to find references.
+        func: The target function to generate source code for.
+        ref_objects: dict of objects referenced by the target function, key is the name and value is the object.
+        classes_to_generate: list of classes which are defined in the same module as the target function.
+            Code generation is required for these classes.
+        generate_code_for_class: Whether the source code for the class shall be generated.
+    """
+    func_module_name = func.__module__
+
+    if generate_code_for_class:
+        # order matters to classes_to_generate, when constructing source code,
+        # referenced classes need to be defined first.
+        classes_to_generate.insert(0, cls)
+
+    inferred_classes: List[Tuple[Any, bool]] = []
+
+    for base_class in cls.__bases__:
+        base_class_name = base_class.__qualname__
+        top_level_class = base_class_name.split(".")[0]
+        if base_class_name == "object":
+            continue
+        elif base_class.__module__ == func_module_name:
+            # if base class is from the same module, we need to parse the class as well as generate code for the class
+            # False in the tuple means this is not a nested class
+            inferred_classes.append((base_class, False))
+        else:
+            # if base class is from another module, we need to import it
+            ref_objects[top_level_class] = func.__globals__[top_level_class]
+
+    # __dict__ contains function, classmethod, classes attributes within a given class which
+    # needs to be further analyzed
+    for v in dict(cls.__dict__).values():
+        if inspect.isclass(v):
+            top_level_cls_name = v.__qualname__.split(".")[0]
+            if v.__module__ == func_module_name:
+                # v is a class defined in the same module as UDF func's, need to dynamically parse the class
+                # if v is class defined in cls, then we should not re-generate code for the nested class
+                inferred_classes.append(
+                    (v, v.__qualname__.startswith(top_level_cls_name))
+                )
+            else:
+                # v is a class defined in another module, import the top level class from another module
+                ref_objects[top_level_cls_name] = v
+        elif inspect.isfunction(v) or isinstance(v, classmethod):
+            # v is a function/classmethod, get the references objects of the Function object
+            get_func_references(
+                v if not isinstance(v, classmethod) else v.__func__, ref_objects
+            )
+        else:
+            # cls.__dict__ would also return __module__, __doc__, __weakref__ which are not required
+            # for code generation, however, class variables is also included in __dict__, we don't do value evaluation
```

Comment:
Will it cause the generated code not working?

Reply:
so far it won't break execution in my test code.

I would say we need to write more test to ensure there's no impact. probably more language reference to understand how python uses them.

we can do more test when I put this feature in the beta branch for source code execution.
it doesn't impact the source code generation.